aspnet/3-binary-data/Controllers/BooksController.cs

```diff
@@ -127,12 +127,18 @@ public ActionResult Edit(long? id)
         // POST: Books/Edit/5
         [HttpPost]
         [ValidateAntiForgeryToken]
-        public ActionResult Edit(Book book, long id)
+        public async Task<ActionResult> Edit(Book book, long id, HttpPostedFileBase image)
         {
             if (ModelState.IsValid)
             {
                 book.Id = id;
                 _store.Update(book);

```

Comment:
On Create we don't know the ID of the book yet (because it hasn't been persisted), so we have to:
 1. Persist the book - `_store.Create(book);`
 2. Upload the image
 3. Update the book in the DB with the ImageUrl

On Edit, we already know the ID of the book so we can simplify this by only persisting once:

```cs
book.Id = id;
if (image != null)
{
    var imageUrl = await _imageUploader.UploadImage(image, book.Id);
    book.ImageUrl = imageUrl;
}
_store.Update(book);
```